From 76699773e186fb2e6e9363499e6426956373e79c Mon Sep 17 00:00:00 2001
From: Selvavignesh Shanmugam <sshanmugam@maxlinear.com>
Date: Tue, 3 Sep 2024 14:50:23 +0800
Subject: [PATCH] WLANRTSYS-84416 Clear SingleLinkAssoc bit for the MultiLink
 STA

Issue:
When ML STA Associates as Single Link and tries to associate again
as MultiLink STA, the previous SingleLinkAssoc bit entry is not
cleared for the current MultiLink Association. Due to this, pruning
of its own Linked STA occurs and ADD_STA & MLD_ADD_STA fails.
This caused TxManager Fatal when Stop Traffic for the main Sta is
sent to the firmware.

Fix:
Clear the SingleLinkAssoc Bit if the current association is multiLink.

FIXUP to WLANRTSYS-62845 MLO changes: MLO Auth Assoc, WPA3 and eapol changes
---
 src/ap/mxl_mld.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/ap/mxl_mld.c b/src/ap/mxl_mld.c
index e8d164a12..e51a6610d 100644
--- a/src/ap/mxl_mld.c
+++ b/src/ap/mxl_mld.c
@@ -1230,6 +1230,9 @@ u16 mxl_multilink_ie_validate(struct hostapd_data *hapd, struct sta_info *sta,
 			memcpy_s(sta->mxl_sta_info.linked_sta->mxl_sta_info.non_ap_mld_mac, ETH_ALEN, sta->mxl_sta_info.non_ap_mld_mac, ETH_ALEN);
 			sta->mxl_sta_info.linked_sta->mxl_sta_info.ext_flags |= WLAN_STA_MULTILINK;
 
+			/* Don't retain the previous SingleLinkAssoc bit entry for the current MultiLink STA*/
+			sta->mxl_sta_info.ext_flags &= ~WLAN_STA_SINGLE_LINK_ML_ASSOC;
+
 			if (!sta->mxl_sta_info.sta_profile_ie)
 				sta->mxl_sta_info.sta_profile_ie = os_zalloc(ie_len);
 			if (!sta->mxl_sta_info.sta_profile_ie) {
-- 
2.43.0

